src/librustc_mir/interpret/operand.rs

```diff
@@ -628,12 +628,7 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                 self.eval_place_to_op(place, layout)?,
 
             Constant(ref constant) => {
-                let layout = from_known_layout(layout, || {
-                    let ty = self.monomorphize(mir_op.ty(self.mir(), *self.tcx), self.substs());
-                    self.layout_of(ty)
-                })?;
-                let op = self.const_value_to_op(constant.literal.val)?;
-                OpTy { op, layout }
+                self.const_to_mplace(constant.literal)?.into()
```

Comment:
You are losing the known layout here, maybe extend `const_to_mplace` to also take an `Option<TyLayout>`?